Add support for ST lsm6dso16is sensor driver

avisconti opened the pull request:
LSM6DSO16IS is a IMU sensor with ISPU capability (https://www.st.com/en/mems-and-sensors/lsm6dso16is.html)                                                                                     
                                                                                     
This PR adds support to accel/gyro data fetching as well as to data fetching of companions sensors attached to the sensor_hub side (i.e. LIS2MDL/HTS221/LPS22HB for the moment)                                                                                    
                                                                                                                                                                         
The driver has been tested using F411 Nucleo board and IKS01A3 shields, with LSM6DSO16IS adapter on the DIL24, and                                                                                  https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/shields/x_nucleo_iks01a3 samples.                                                                       
                                                                                    
EDIT:                                                                                    
Follow same approach as #57517 in sensor hub for inclusive naming

avisconti committed:
drivers/sensor: add support to LSM6DSO16IS IMU sensor

The LSM6DSO16IS is a system-in-package featuring a 3-axis digital
accelerometer and a 3-axis digital gyroscope for industrial and IoT
solutions. The LSM6DSO16IS embeds a new ST category of processing,
ISPU (intelligent sensor processing unit) to support real-time applications
that rely on sensor data. The ISPU is an ultra-low-power, high-performance
programmable core which can execute signal processing and AI algorithms
in the edge.

https://www.st.com/en/mems-and-sensors/lsm6dso16is.html

This driver is based on stmemsc HAL i/f v2.02

Signed-off-by: Armando Visconti <armando.visconti@st.com>

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_own.conf:
```diff
@@ -32,6 +32,7 @@ CONFIG_LIS3MDL_TRIGGER_OWN_THREAD=y
 CONFIG_LPS22HH_TRIGGER_OWN_THREAD=y
 CONFIG_LSM6DSL_TRIGGER_OWN_THREAD=y
 CONFIG_LSM6DSO_TRIGGER_OWN_THREAD=y
+CONFIG_LSM6DSO16IS_TRIGGER_OWN_THREAD=y
```
Add this sensor to `sensors_trigger_global.conf` and `sensors_trigger_none.conf` too please

avisconti replied:
> Add this sensor to `sensors_trigger_global.conf` and `sensors_trigger_none.conf` too please

yep! Thanks for catching it...

MaureenHelm reviewed (commented):
This part looks pretty cool. When can we run Zephyr on the ISPU? :)

avisconti commented:
> This part looks pretty cool. When can we run Zephyr on the ISPU? :)

It might be an interesting exercise... :)

avisconti committed:
tests: drivers: build_all: add sensorhub test case

Add test case for all drivers that are supporting the sensorhub
feature, and fix right away all the yet undiscovered issues.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
I rebased and solved the conflict on ```tests/drivers/build_all/sensor/i2c.dtsi```. Moreover I added a commit for testing all sensor sensorhub cases, and I fixed few issues that came out from this test..

avisconti committed:
drivers/sensor: stmemsc: add macros to populate stmdev_ctx_t structure

Add STMEMSC_CTX I2C/SPI/I3C macros that can be reused among all STMEMSC
based ST sensor drivers to populate the stmdev_ctx_t stucture.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

erwango commented:
@avisconti Another rebase is needed.

MaureenHelm commented on drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c:
```diff
@@ -0,0 +1,838 @@
+/*
+ * use SLV0 for generic read to slave device
+ */
+static int lsm6dso16is_shub_read_slave_reg(const struct device *dev,
```
Use an alternative to the word "slave"
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

MaureenHelm reviewed (changes requested):
Everything else looks good

avisconti commented on drivers/sensor/lsm6dso16is/lsm6dso16is.c:
```diff
@@ -0,0 +1,933 @@
+	}
+
+	/* Wait sensor turn-on time as per datasheet */
+	k_busy_wait(35 * USEC_PER_MSEC);
```
ooops, sure... :(

avisconti commented:
"tests/drivers/build_all/sensor/i2c.dtsi"  seems to be the collector of everyone's contribution... @MaureenHelm @erwango @teburd if you please can spend some time to dig into this PR I appreciate. :D

MaureenHelm commented on drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c:
```diff
@@ -0,0 +1,838 @@
+
+#define LSM6DSO16IS_SHUB_DATA_OUT				0x02
+
+#define LSM6DSO16IS_SHUB_SLV0_ADDR				0x15
+#define LSM6DSO16IS_SHUB_SLV0_SUBADDR			0x16
+#define LSM6DSO16IS_SHUB_SLV0_CONFIG			0x17
+#define LSM6DSO16IS_SHUB_SLV1_ADDR				0x18
+#define LSM6DSO16IS_SHUB_SLV1_SUBADDR			0x19
+#define LSM6DSO16IS_SHUB_SLV1_CONFIG			0x1A
+#define LSM6DSO16IS_SHUB_SLV2_ADDR				0x1B
+#define LSM6DSO16IS_SHUB_SLV2_SUBADDR			0x1C
+#define LSM6DSO16IS_SHUB_SLV2_CONFIG			0x1D
+#define LSM6DSO16IS_SHUB_SLV3_ADDR				0x1E
+#define LSM6DSO16IS_SHUB_SLV3_SUBADDR			0x1F
+#define LSM6DSO16IS_SHUB_SLV3_CONFIG			0x20
+#define LSM6DSO16IS_SHUB_SLV0_DATAWRITE			0x21
```
Replace "SLV" too

avisconti replied:
I could delete those names and use the ones defined in https://github.com/zephyrproject-rtos/hal_st/blob/master/sensor/stmemsc/lsm6dsv16x_STdC/driver/lsm6dsv16x_reg.h, like LSM6DSV16X_SLV0_ADD. Still the SLV keyword in it, but it's a h/w register name. I'll do the change anyway, since it is cleaner...

teburd replied:
If these are the documented register names it probably doesn't make sense to change. This is in a .c file not in a dts binding, kconfig, or header than an application may directly interact with. So probably ok in my mind to leave as is, but would be nice if st similarly updated their naming scheme.

avisconti replied:
> If these are the documented register names it probably doesn't make sense to change. It would make it difficult to correlate the defines to the docs. This is in a .c file not in a dts binding, kconfig, or header than an application may directly interact with. So probably ok in my mind to leave as is, but would be nice if st similarly updated their naming scheme.

Yes, I will keep using the h/w reg names. Still, a change here is requested as it would be better to use directly the names defined in STMEMSC HAL and not redefine them. And yes, STM is going to uniform the names for the future devices (not the current/past ones) with the inclusive naming schemes.

avisconti replied:
For the rest of names, I will replace ```slave``` with ```target``` (in all the possible declinations). Thanks!

MaureenHelm commented on drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c:
```diff
@@ -0,0 +1,838 @@
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t device[3];
+
+	device[0] = (slv_addr << 1) | LSM6DSO16IS_SHUB_SLVX_READ;
```
Replace "slv"

MaureenHelm commented on drivers/sensor/lsm6dso16is/lsm6dso16is_shub.c:
```diff
@@ -0,0 +1,838 @@
+}
+
+/*
+ * SLAVEs configurations:
+ *
+ *  - SLAVE 0: used for configuring all devices
+ *  - SLAVE 1: used as data read channel for external device #1
+ *  - SLAVE 2: used as data read channel for external device #2
+ *  - SLAVE 3: used for generic reads while data channel is enabled
```
slave

avisconti replied:
yes, master! (hope I can joke a bit about it... :D ) OK, I'll take care of it. Sorry, I just looked for the word "slave" in lowe case.

MaureenHelm reviewed (changes requested):
> "tests/drivers/build_all/sensor/i2c.dtsi" seems to be the collector of everyone's contribution... @MaureenHelm @erwango @teburd if you please can spend some time to dig into this PR I appreciate. :D

Sorry, I know it's a pain. @yperess was going to look at alternatives. I wonder if it's possible to generate `tests/drivers/build_all/sensor/i2c.dtsi` from the available sensor bindings...

avisconti commented:
Repushed following same approach as in #57517 (for lsm6dso sensor hub)

avisconti commented:
@MaureenHelm @teburd @erwango Can you possibly take a look at this PR? I am continuosly solving the conflicts for the sensor build_all test :D Thanks
